flink/src/main/java/org/apache/iceberg/flink/IcebergTableSource.java

```diff
@@ -100,70 +93,73 @@ public boolean isBounded() {
         .build();
   }
 
-  @Override
-  public TableSchema getTableSchema() {
-    return schema;
-  }
-
-  @Override
-  public DataType getProducedDataType() {
-    return getProjectedSchema().toRowDataType().bridgedTo(RowData.class);
-  }
-
   private TableSchema getProjectedSchema() {
-    TableSchema fullSchema = getTableSchema();
     if (projectedFields == null) {
-      return fullSchema;
+      return schema;
     } else {
-      String[] fullNames = fullSchema.getFieldNames();
-      DataType[] fullTypes = fullSchema.getFieldDataTypes();
+      String[] fullNames = schema.getFieldNames();
+      DataType[] fullTypes = schema.getFieldDataTypes();
       return TableSchema.builder().fields(
           Arrays.stream(projectedFields).mapToObj(i -> fullNames[i]).toArray(String[]::new),
           Arrays.stream(projectedFields).mapToObj(i -> fullTypes[i]).toArray(DataType[]::new)).build();
     }
   }
 
   @Override
-  public String explainSource() {
-    String explain = "Iceberg table: " + loader.toString();
-    if (projectedFields != null) {
-      explain += ", ProjectedFields: " + Arrays.toString(projectedFields);
-    }
-
-    if (isLimitPushDown) {
-      explain += String.format(", LimitPushDown : %d", limit);
-    }
+  public void applyLimit(long newLimit) {
+    this.limit = newLimit;
+  }
 
-    if (isFilterPushedDown()) {
-      explain += String.format(", FilterPushDown: %s", COMMA.join(filters));
+  @Override
+  public Result applyFilters(List<ResolvedExpression> flinkFilters) {
+    List<ResolvedExpression> acceptedFilters = Lists.newArrayList();
+    List<Expression> expressions = Lists.newArrayList();
+
+    for (ResolvedExpression resolvedExpression : flinkFilters) {
+      Optional<Expression> icebergExpression = FlinkFilters.convert(resolvedExpression);
+      if (icebergExpression.isPresent()) {
+        expressions.add(icebergExpression.get());
+        acceptedFilters.add(resolvedExpression);
+      }
     }
 
-    return TableConnectorUtils.generateRuntimeName(getClass(), getTableSchema().getFieldNames()) + explain;
+    this.filters = expressions;
+    return Result.of(acceptedFilters, flinkFilters);
```

Comment:
I read the  javadoc of [SupportsFilterPushDown](https://github.com/apache/flink/blob/97bfd049951f8d52a2e0aed14265074c4255ead0/flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsFilterPushDown.java#L63),  the remanningFilters don't have to be the full list of origin `flinkFilters`,   it could be the list of filters that did  push down for iceberg source.

Although it will be surely correct if pass the complete list as `remainingFilters`, but that seems will introduce extra resources to apply the `acceptedFilters` twice.

Reply:
yes , I also think the remainingFilters is the filters which could not be applied by the source.
but when I do that, there were a lot of errors in `TestFlinkTableSource` .
So I change the `remainingFilters` to origin `flinkFilters `

Reply:
What's the reason that fails the unit tests ? I think we need to have a clear understanding about the semantics of `acceptedFilters` and `remainingFilters` before we change the code to make unit tests pass.

Reply:
I tried to apply the following patch, all unit tests could be passed: 

```diff
diff --git a/flink/src/main/java/org/apache/iceberg/flink/IcebergTableSource.java b/flink/src/main/java/org/apache/iceberg/flink/IcebergTableSource.java
index 79cb17c22..1c2df292b 100644
--- a/flink/src/main/java/org/apache/iceberg/flink/IcebergTableSource.java
+++ b/flink/src/main/java/org/apache/iceberg/flink/IcebergTableSource.java
@@ -124,17 +124,20 @@ public class IcebergTableSource
   @Override
   public Result applyFilters(List<ResolvedExpression> flinkFilters) {
     List<ResolvedExpression> acceptedFilters = Lists.newArrayList();
-    List<Expression> expressions = Lists.newArrayList();
-
-    for (ResolvedExpression resolvedExpression : flinkFilters) {
-      Optional<Expression> icebergExpression = FlinkFilters.convert(resolvedExpression);
-      if (icebergExpression.isPresent()) {
-        expressions.add(icebergExpression.get());
-        acceptedFilters.add(resolvedExpression);
+    List<ResolvedExpression> remainingFilters = Lists.newArrayList();
+    List<Expression> acceptedExpressions = Lists.newArrayList();
+
+    for (ResolvedExpression flinkFilter : flinkFilters) {
+      Optional<Expression> pushedDownFilter = FlinkFilters.convert(flinkFilter);
+      if (pushedDownFilter.isPresent()) {
+        acceptedFilters.add(flinkFilter);
+        acceptedExpressions.add(pushedDownFilter.get());
+      } else {
+        remainingFilters.add(flinkFilter);
       }
     }
 
-    this.filters = expressions;
+    this.filters = acceptedExpressions;
     return Result.of(acceptedFilters, flinkFilters);
   }
```

Reply:
OK, after read the flink runtime & iceberg filter pushdown code carefully,  I found that we iceberg's AVRO file reader  did not push down the filters, so it will always return the row that are not filtered by `acceptedFilters` .

![image](https://user-images.githubusercontent.com/5028729/110607435-276e9b80-81c6-11eb-836e-d89fb11a8aba.png)

So here we could only pass the `flinkFilters` as the `remainingFilters`.  Let me resolve this comment.